compiler/rustc_mir_transform/src/streamline_iter.rs

```diff
@@ -0,0 +1,366 @@
+            let pointer_assign = Rvalue::Use(Operand::Copy(pointer));
+            let pointer_assign = Statement {
+                kind: StatementKind::Assign(Box::new((pointer_local.into(), pointer_assign))),
+                source_info,
+            };
+            let borrow = if *muta == Mutability::Not {
+                BorrowKind::Shared
+            } else {
+                BorrowKind::Mut { kind: MutBorrowKind::Default }
+            };
+            let rf = Rvalue::Ref(*region, borrow, tcx.mk_place_deref(pointer_local.into()));
+            let rf = Statement {
+                kind: StatementKind::Assign(Box::new((ref_local.into(), rf))),
+                source_info,
+            };
+            let some_val = Rvalue::Aggregate(
+                Box::new(AggregateKind::Adt(
+                    option_adt.did(),
+                    VariantIdx::from_usize(1),
+                    option_gargs,
+                    None,
+                    None,
+                )),
+                [Operand::Move(ref_local.into())].into(),
+            );
+            let set_some = Statement {
+                kind: StatementKind::Assign(Box::new((destination, some_val))),
+                source_info,
+            };
+            some_block.statements.push(pointer_assign);
+            some_block.statements.push(rf);
+            some_block.statements.push(incr);
+            some_block.statements.push(set_some);
+
+            // Get the new blocks in place - this invalidates caches!
+            cfg_invalid = true;
+            let some_bb = bbs.push(some_block);
+            let none_bb = bbs.push(none_block);
+
+            // Change the original block.
+            let mut bb = &mut bbs[bid];
+            bb.terminator = Some(Terminator {
+                kind: TerminatorKind::SwitchInt {
+                    discr: Operand::Move(check.into()),
+                    targets: SwitchTargets::new(std::iter::once((0, some_bb)), none_bb),
+                },
+                source_info,
+            });
+            bb.statements.push(ptr_cast);
+            bb.statements.push(check_iter_empty);
+        }
+        if cfg_invalid {
+            body.basic_blocks.invalidate_cfg_cache();
+        }
+    }
+
+    fn is_required(&self) -> bool {
+        true
+    }
+}
+fn not_zst<'tcx>(t: Ty<'tcx>, tcx: TyCtxt<'tcx>) -> bool {
+    match t.kind() {
+        ty::Uint(_)
+        | ty::Int(_)
+        | ty::Bool
+        | ty::Float(_)
+        | ty::Char
+        | ty::Ref(..)
+        | ty::RawPtr(..)
+        | ty::FnPtr(..) => true,
+        ty::Tuple(elements) => elements.iter().any(|ty| not_zst(ty, tcx)),
+        ty::Array(elem, count) if count.try_to_target_usize(tcx).is_some_and(|count| count > 0) => {
+            not_zst(*elem, tcx)
+        }
+        ty::Array(_, _) => false,
+        ty::Never | ty::FnDef(..) => false,
+        ty::Adt(def, args) => match def.adt_kind() {
+            AdtKind::Enum => def.variants().len() > 1,
```

Comment:
Hehehe, very familiar function... I have made this mistake too, this will blow up for enums with uninhabited variants.

If it helps you can steal from here: https://github.com/rust-lang/rust/pull/116542/commits/2c19fc66327109a75e6b6602c95aecf93e7fc30e#diff-d5c6f94594e21dda2a9e3717a8b245c481fd461d8eece9c7bc960693cb0c368aR795